bootstore/src/schemes/v0/peer_networking.rs

```diff
@@ -0,0 +1,635 @@
+            write_sock,
+            read_sock,
+            main_tx.clone(),
+            rx,
+            log.clone(),
+        );
+
+        // We always exit on server tasks, as the remote peer
+        // will reconnect.
+        let _ = conn.run().await;
+    });
+
+    AcceptedConnHandle { handle, tx, addr, unique_id }
+}
+
+// Serialize and write `msg` into `buf`, prefixed by a 4-byte big-endian size header
+//
+// Return the total amount of data written into `buf` including the 4-byte header
+fn write_framed<T: Serialize + ?Sized>(
+    msg: &T,
+) -> Result<Vec<u8>, ciborium::ser::Error<std::io::Error>> {
+    let mut cursor = Cursor::new(vec![]);
+    // Write a size placeholder
+    std::io::Write::write(&mut cursor, &[0u8; FRAME_HEADER_SIZE]);
+    cursor.set_position(FRAME_HEADER_SIZE as u64);
+    ciborium::into_writer(msg, &mut cursor)?;
+    let size: u32 =
+        (cursor.position() - FRAME_HEADER_SIZE as u64).try_into().unwrap();
+    let mut buf = cursor.into_inner();
+    buf[0..FRAME_HEADER_SIZE].copy_from_slice(&size.to_be_bytes());
+    Ok(buf)
+}
+
+// Decode the 4-byte big-endian frame size header
+fn read_frame_size(buf: [u8; FRAME_HEADER_SIZE]) -> usize {
+    u32::from_be_bytes(buf) as usize
+}
+
+#[derive(Debug, From)]
+enum HandshakeError {
+    Serialization(ciborium::ser::Error<std::io::Error>),
+    Deserialization(ciborium::de::Error<std::io::Error>),
+    Io(tokio::io::Error),
+    UnsupportedScheme,
+    UnsupportedVersion,
+    Timeout,
+}
+
+// Perform scheme/version negotiation and exchange peer_ids for scheme v0
+async fn perform_handshake(
+    sock: TcpStream,
+    local_peer_id: &Baseboard,
+    local_addr: SocketAddrV6,
+    log: &Logger,
+) -> Result<(OwnedReadHalf, OwnedWriteHalf, Identify), HandshakeError> {
+    // Enough to hold the `Hello` and `Identify` messages
+    let mut read_buf = [0u8; 128];
+    let (mut read_sock, mut write_sock) = sock.into_split();
+
+    // Serialize and write the handshake messages into `write_buf`
+    let mut hello_cursor = Cursor::new(Hello::default().serialize());
+    let identify = write_framed(&Identify {
+        id: local_peer_id.clone(),
+        addr: local_addr,
+    })?;
+    let mut identify_cursor = Cursor::new(&identify);
+
+    let handshake_start = Instant::now();
+
+    // Read `Hello` and the frame size of `Identify`
+    let initial_read = Hello::serialized_size() + FRAME_HEADER_SIZE;
+
+    let mut total_read = 0;
+    let mut identify_len = 0;
+    let mut identify: Option<Identify> = None;
+
+    loop {
+        let timeout =
+            KEEPALIVE_TIMEOUT.saturating_sub(Instant::now() - handshake_start);
+
+        let end = initial_read + identify_len;
+
+        let hello_written = !hello_cursor.has_remaining();
+
+        if identify.is_some() && !identify_cursor.has_remaining() {
+            return Ok((read_sock, write_sock, identify.unwrap()));
+        }
+
+        tokio::select! {
+            _ = sleep(timeout) => {
```

Comment:
To confirm, the timeout applies to every loop iteration, not the entire operation. Is this desired? (In many cases it is, just confirming.)

Reply:
Yeah, its per iteration. I adjust it at the top of the loop to account for time moving on.